src/lib.rs

```diff
@@ -258,31 +280,186 @@ pub struct LineSender {
     last_line_start: usize
 }
 
-impl LineSender {
-    pub fn connect(host: &str, port: &str, net_interface: Option<&str>) -> Result<Self> {
-        let addr: SockAddr = gai::resolve_host_port(host, port)?;
+#[derive(Debug, Clone)]
+pub struct AuthParams<'a> {
+    key_id: &'a str,
+    priv_key: &'a str,
+    pub_key_x: &'a str,
+    pub_key_y: &'a str
+}
+
+#[derive(Debug, Clone)]
+pub struct LineSenderBuilder<'a> {
+    capacity: usize,
+    host: &'a str,
+    port: &'a str,
+    net_interface: Option<&'a str>,
+    auth: Option<AuthParams<'a>>
+}
+
+impl <'a> LineSenderBuilder<'a> {
+    /// QuestDB server and port.
+    pub fn new(host: &'a str, port: &'a str) -> Self {
+        Self {
+            capacity: 65536,
+            host: host,
+            port: port,
+            net_interface: None,
+            auth: None }
+    }
+
+    /// Set the initial buffer capacity.
+    pub fn capacity(&mut self, byte_count: usize) -> &mut Self {
+        self.capacity = byte_count;
+        self
+    }
+
+    /// Select local outbound interface.
+    pub fn net_interface(&mut self, addr: &'a str) -> &mut Self {
+        self.net_interface = Some(addr);
+        self
+    }
+
+    /// Authentication Parameters.
+    pub fn auth(
+        &mut self, key_id: &'a str,
+        priv_key: &'a str,
+        pub_key_x: &'a str,
+        pub_key_y: &'a str) -> &mut Self
+    {
+        self.auth = Some(AuthParams {
+            key_id: key_id,
+            priv_key: priv_key,
+            pub_key_x: pub_key_x,
+            pub_key_y: pub_key_y
+        });
+        self
+    }
+
+    /// Connect synchronously.
+    pub fn connect(self) -> Result<LineSender> {
+        let addr: SockAddr = gai::resolve_host_port(self.host, self.port)?;
         let sock = Socket::new(Domain::IPV4, Type::STREAM, Some(Protocol::TCP))
-            .map_err(|io_err| map_io_to_socket_err("Could not open TCP socket: ", io_err))?;
+            .map_err(|io_err| map_io_to_socket_err(
+                "Could not open TCP socket: ", io_err))?;
         sock.set_nodelay(true)
```

Comment:
nit: is there a particular reason to disable naggling in ILP clients? I'm just curious.

Reply:
Not sure I follow you: What do you mean by "disable nagging"?

You might have read `map_err` as consuming the error and throwing it away. That is not the case:
A bit of Rust background on error handling and an explaination of what's happening here:
* Functions in Rust return things that can go wrong as `Result<Thing, SomeError>`.
* This is a sum type that should be read roughly as "Either a `Thing` or a `SomeError` error.".
* My Rust API returns a single consistent error type.
* Here `.map_err` takes a function that can transform the error whilst keeping the value untouched.
* In other words, `.map_err` transforms any error and does not throw it away, i.e. maps `Result<T, E1>` to `Result<T, E2>`.
* The `?` is a short-hand notation that inserts an `if` statement and early-returns in case of errors.
 In other words, if you had a `let x: Result<T, E> = ...`, the type of the expression `x?` is simply `T`.

Reply:
Ah, sorry: I never heard Nagle's algorithm referred as naggling.

I thought you were commenting on the error handling! 

The reason why nodelay is set comes from conversations with Vlad where he argued it should be set. 

In short, if there's a pending line which is small, we don't want to delay it. 

Of there's a lot of buffered data, the nodelay does nothing, so there's no harm keeping it.

Reply:
Done. No-op.

Reply:
> Also note that without nodelay, flush wouldn't actually flush.

The main promise of `write()` system call is to copy the data into OS buffer. All cases when naggling was disabled I saw in OSS projects are related with small enough messages and, usually, per key/row operations. In our case, writes are done in batches, so I would be surprised if disabling Nagle's algorithm really changes anything. Nevertheless, let's keep it as is. I was just curious if there is a particular reason to disable it here.